### ISSUE TITLE
DCML>RNTXT conversion (bug). Metric distortions due to empty labels

### ISSUE THREAD

**Issue by johentsch** [MarkGotham/When-in-Rome#82]
Hi @MarkGotham and @malcolmsailor,

The m21 stream of Tchaikovsky op37a08 has a metric bump in m. 5 whose offset comes an eighth early:

```
{0.0} <music21.metadata.Metadata object at 0x7f402fad6d40>
{0.0} <music21.stream.Part 0x7f40dccc9d20>
    {0.0} <music21.stream.Measure 0 offset=0.0>
        {0.0} <music21.key.Key of b minor>
        {0.0} <music21.meter.TimeSignature 6/8>
        {0.0} <music21.roman.RomanNumeral i6 in b minor>
    {1.0} <music21.stream.Measure 1 offset=1.0>
        {0.0} <music21.roman.RomanNumeral i in b minor>
        {1.0} <music21.roman.RomanNumeral #viio6 in b minor>
        {2.0} <music21.roman.RomanNumeral VI6 in b minor>
    {4.0} <music21.stream.Measure 2 offset=4.0>
        {0.0} <music21.roman.RomanNumeral VI6 in b minor>
        {1.0} <music21.roman.RomanNumeral V6 in b minor>
        {2.0} <music21.roman.RomanNumeral vo6 in b minor>
    {7.0} <music21.stream.Measure 3 offset=7.0>
        {0.0} <music21.roman.RomanNumeral vo6 in b minor>
        {1.0} <music21.roman.RomanNumeral IV6 in b minor>
        {2.0} <music21.roman.RomanNumeral #viio2 in b minor>
    {10.0} <music21.stream.Measure 4 offset=10.0>
        {0.0} <music21.roman.RomanNumeral V[no5][add6] in b minor>
        {1.0} <music21.roman.RomanNumeral #viio64 in b minor>
        {1.5} <music21.roman.RomanNumeral i in b minor>
    {12.5} <music21.stream.Measure 5 offset=12.5>
        {0.0} <music21.roman.RomanNumeral VI in b minor>
        {1.0} <music21.roman.RomanNumeral iv65 in b minor>
        {1.5} <music21.roman.RomanNumeral Ger6 in b minor>
```

Inspecting the rntext reveals an empty label at the end of m. 4:

https://github.com/MarkGotham/When-in-Rome/blob/30b3240d8ac58b72b417cbd37523b40486e9606b/Corpus/Keyboard_Other/Tchaikovsky%2C_Pyotr/Seasons%2C_Op.37a/8/analysis.txt#L11-L13

The problem is caused by non-harmony labels such as `{` and seems to be happening systematically, e.g.:

https://github.com/MarkGotham/When-in-Rome/blob/30b3240d8ac58b72b417cbd37523b40486e9606b/Corpus/Keyboard_Other/Tchaikovsky%2C_Pyotr/Seasons%2C_Op.37a/7/analysis.txt#L24

The problem can easily be avoided by skipping all rows that have an empty value in the column `chord` while converting, e.g. 

https://github.com/DCMLab/tchaikovsky_seasons/blob/b1cd416d972cca31349d348bc8383f2e16d73269/harmonies/op37a07.tsv#L45

**Comment by MarkGotham** [comment]
Thanks @johentsch for clear and detailed bug report!
@malcolmsailor will you field this one over on music21, please?

**Comment by johentsch** [comment]
I realize that m21 would be the better place to file this issue. For this it would be amazing to see metadata that refer to the conversion functions/methods and the music21 version at work. Maybe I've just missed it...

**Comment by MarkGotham** [comment]
Thanks. We can move over there when ready ... but it's good to get a specific proposal worked up before trying to breach the gates of music21 ;)

We recently settled on declaring an updated version of the DCML-romantext parser and could include that as well as yours on each conversion file (i.e., `DCML v2` and `DCML-romantext parser v. 2023.02` or similar).

Do you also imagine including the music21 version? At a quick look, this seems to be a / the way to retrieve it at runtime:
```
>>> from music21 import _version
>>> _version.__version_info__
('9', '0', '0', 'a6')
```

**Comment by johentsch** [comment]
Yes, I started looking into parsing rntxt for enabling its use in our analysis library and would like to make it a point to be very verbose on this kind of details when storing outputs. Currently I'm doing stuff like

```python
import ms3
from git import Repo

data_repository = Repo(corpus_path)
print(f"data_repository @ {data_repository.commit().hexsha[:7]}")
print(f"ms3 version {ms3.__version__}") # should work with most libraries
```

Ideally, we might even go to the level of the concrete commands/parameters invoked. ms3 will need to improve in this regard, too...

**Comment by MarkGotham** [comment]
Thanks @johentsch .

@malcolmsailor [this commit](941ec9b39e756c6ae8c83cdd48e771ef34f9683b) is a "warts and all" with several of these metric distortions included.

E.g., `Piano_Sonatas/Mozart,_Wolfgang_Amadeus/K282/3/`.

To replicate, run `python3 -m Code.collect_convert --update_all_DCML_analyses_from_json` 

Thanks for all / any help.

**Comment by malcolmsailor** [comment]
> I realize that m21 would be the better place to file this issue. For this it would be amazing to see metadata that refer to the conversion functions/methods and the music21 version at work. Maybe I've just missed it...

@johentsch what do you mean exactly? Metadata in the file header? I think maybe this request should take a separate issue at music21?

**Comment by MarkGotham** [comment]
Classic. This issue has become a double issue, leading to predictable confusion.

- This issue as per the title "metric distortions ..." (medium priority)
- [New issue for metadata here](https://github.com/MarkGotham/When-in-Rome/issues/90) (low priority). I'm deleting comments about metadata from here and summarising them over there.

Both are really issues for music21 but I'm happy to use this space for working up a proposal if that helps.
